travis/setup_dependencies_common.sh

```diff
@@ -313,7 +374,85 @@ if [[ ! -z $CONDA_DEPENDENCIES ]]; then
     fi
 fi
 
+if [[ ! -z $CONDA_DEPENDENCIES ]]; then
+    # Do a dry run of the conda install here to make sure that pins are
+    # ACTUALLY being respected. This will become unnecessary when
+    # https://github.com/conda/conda/issues/9052
+    # is fixed
+
+    # NOTE: it is important that the expression below remain in an if context
+    # because of the 'set -e' above, which causes the shell to immediately
+    # exit if the last command in a pipeline has a non-zero exit status,
+    # UNLESS the pipeline is in a few specific contexts.  From
+    # https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html:
+    #
+    #     The shell does not exit if the command that fails is ...part of the
+    #     test in an if statement...
+    #
+    # Use tee to print output to console and to file to avoid travis timing out
+    _tmp_output_file="tmp.txt"
+    # do not exit on failure of the dry run because pip fallback may succeed
+    set +e
+
+    conda install --dry-run $CONDA_DEPENDENCIES > >(tee $_tmp_output_file) 2>&1
```

Comment:
Out of curiosity what does `> >` mean?

Reply:
I know so much more arcane bash now than I did before; this bit of it is thanks to @djhoese. It turns out that `>(command)` pipes stdout into `command`; its partner `<(command)` connects the output of command to stdin. 

At least I think I have that right....

Reply:
So this is known as "process substitution" iirc and I only found it because of trying to work on this issue. The idea is that most flavors of bash/sh will put any piped commands (ex. `one_cmd | two_cmd | three_cmd | tee tmp_file`) in a subshell. Normally, for most bash scripts, this isn't a huge deal, but for conda it is because the `conda` shell command is modifying `PATH` and running other scripts so that they effect our current environment. If they were run in a subshell then none of those changes take effect.

So what is happening is the `conda install` command is running in the current shell and then piping its output to a new process (which is running `tee`).

The other option would be to deactivate and reactivate conda every time we run a conda command in these setup scripts.